opentelemetry-sdk/src/metrics/internal/histogram.rs

```diff
@@ -54,30 +25,37 @@ struct Buckets<T> {
     max: T,
 }
 
-impl<T: Number> Buckets<T> {
-    /// returns buckets with `n` bins.
-    fn new(n: usize) -> Buckets<T> {
-        Buckets {
-            counts: vec![0; n],
+impl<T> Buckets<T>
+where
+    T: Number,
+{
+    fn new(size: usize) -> Self {
+        Self {
+            counts: vec![0; size],
             min: T::max(),
             max: T::min(),
             ..Default::default()
         }
     }
-
-    fn sum(&mut self, value: T) {
-        self.total += value;
-    }
-
-    fn bin(&mut self, idx: usize, value: T) {
+    fn update(&mut self, config: &BucketsConfig, f_value: f64, measurement: T) {
+        // This search will return an index in the range `[0, bounds.len()]`, where
+        // it will return `bounds.len()` if value is greater than the last element
+        // of `bounds`. This aligns with the buckets in that the length of buckets
+        // is `bounds.len()+1`, with the last bucket representing:
+        // `(bounds[bounds.len()-1], +∞)`.
+        let idx = config.bounds.partition_point(|&x| x < f_value);
```

Comment:
The calculation of index should happen outside of the lock of `Mutex<Buckets<T>>`. You could do it before calling `Buckets<T>::update()` method.

Reply:
There would be a perf regression in throughput (measured by the [stress tests](https://github.com/open-telemetry/opentelemetry-rust/blob/main/stress/src/metrics_histogram.rs)) with this refactoring. Even if you move the index calculation outside of the `Mutex` lock, you would still be calculating it under the `RwLock` of ValueMap. Currently, we do it outside of `ValueMap` so threads don't even have to contend even for the read lock to perform this operation.

@fraillt Would it be possible to calculate the index outside of any locks and pass it to the `ValueMap` update methods?